### PR TITLE
[storage] get_rightmost_leaf take a version

### PR DESCRIPTION
**Pull request by lightmark**
### Description

Take a version now so fast sync wouldn't take a wrong leaf node at a different version in any case.

### Test Plan
ut

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/4123)

### CODE REVIEW

File: storage/aptosdb/src/state_merkle_db.rs [hunk 1]
```diff
@@ -257,17 +263,17 @@ impl TreeReader<StateKey> for StateMerkleDb {
         Ok(node_opt)
     }
 
-    fn get_rightmost_leaf(&self) -> Result<Option<(NodeKey, LeafNode)>> {
+    fn get_rightmost_leaf(&self, version: Version) -> Result<Option<(NodeKey, LeafNode)>> {
         // Since everything has the same version during restore, we seek to the first node and get
         // its version.
         let mut iter = self.iter::<JellyfishMerkleNodeSchema>(Default::default())?;
-        iter.seek_to_first();
+        iter.seek(&(version, 0))?;
```
Comment (msmouse) [inline storage/aptosdb/src/state_merkle_db.rs:270]: Don't need to re-assign to version any more.

File: storage/aptosdb/src/state_merkle_db.rs [hunk 2]
```diff
@@ -257,17 +263,17 @@ impl TreeReader<StateKey> for StateMerkleDb {
...
         let version = match iter.next().transpose()? {
```
Comment (msmouse) [inline storage/aptosdb/src/state_merkle_db.rs:271]: Don't need to re-assign to version any more.

File: storage/aptosdb/src/state_store/mod.rs [hunk 3]
```diff
@@ -715,11 +715,12 @@ impl StateStore {
         version: Version,
         expected_root_hash: HashValue,
     ) -> Result<Box<dyn StateSnapshotReceiver<StateKey, StateValue>>> {
-        Ok(Box::new(StateSnapshotRestore::new_overwrite(
+        Ok(Box::new(StateSnapshotRestore::new(
             &self.state_merkle_db,
             self,
             version,
             expected_root_hash,
+            true, /* async_commit */
```
Comment (msmouse) [inline storage/aptosdb/src/state_store/mod.rs:723]: Use false and comment: 
```
// state sync doesn't query for the progress, but keeps its record by itself. 
// TODO: change to async comment once it does like https://github.com/aptos-labs/aptos-core/blob/159b00f3d53e4327523052c1b99dd9889bf13b03/storage/backup/backup-cli/src/backup_types/state_snapshot/restore.rs#L147 or overlap at least two chunks.
```

Comment (msmouse) [reply storage/aptosdb/src/state_store/mod.rs:723]: @JoshLind

**Review by msmouse — APPROVED** [review]
I tested fast syncing a fullnode with interruptions, and it looks working.

**Review by grao1991 — APPROVED** [review]

**Review by zcchahaha — APPROVED** [review]

**Pull request merged**